docs/design/tools/illink/feature-switch-attributes.md

```diff
@@ -0,0 +1,679 @@
+
+  On its own this is not enough for ILLink/ILCompiler to do branch elimination, because there's no tie to the feature switch name.
+
+- a reference to the feature name string:
+
+  ```csharp
+  class Feature {
+      [FeatureGuard("System.Runtime.CompilerServices.RuntimeFeature.IsDynamicCodeSupported")]
+      public static bool IsSupported => RuntimeFeature.IsDynamicCodeSupported;
+  }
+  ```
+
+  The analyzer would need to hard-code the fact that this string corresponds to `RequiresDynamicCodeAttribute`, unless we model this relationship via attributes that represent feature switches.
+
+  This would be sufficient for ILLink/ILCompiler to treat `IsSupported` as a constant based on the feature switch name.
+
+- a reference to the existing feature check property:
+
+  ```csharp
+  class Feature {
+      [FeatureGuard(typeof(RuntimeFeature), nameof(RuntimeFeature.IsDynamicCodeSupported))]
+      public static bool IsSupported => RuntimeFeature.IsDynamicCodeSupported;
+  }
+  ```
+
+  The analyzer would need to hard-code the fact that `RuntimeFeature.IsDynamicCodeSupported` corresponds to `RequiresDynamicCodeAttribute`, unless we model this relationship via attributes that represent feature switches.
+
+  This would be sufficient for ILLink/ILCompiler to treat `IsSupported` as a constant based on the feature switch name, assuming it has existing knowledge of the fact that `RuntimeFeature.IsDynamicCodeSupported` is contrelled by the feature switch named `"System.Runtime.CompilerServices.RuntimeFeature.IsDynamicCodeSupported"`, either from a substitution XML or from a separate attribute that encodes this.
+
+### Danger of incorrect usage
+
+Since a feature guard silences warnings from the analyzer, there is some danger that `FeatureGuardAttribute` will be used carelessly as a way to silence warnings, even when the definition of the `IsSupported` property doesn't have any tie to the existing feature. For example:
+
+```csharp
+class Feature {
+    [FeatureGuard(typeof(RequiresDynamicCodeAttribute))]
+    public static bool SilenceWarnings => true; // BAD
+}
+```
+
+```csharp
+if (Feature.SilenceWarnings) {
+    APIWhichRequiresDynamicCode(); // No warnings
+}
+```
+
+This will silence warnings from the analyzer without any indication that `RequiresDynamicCode`-annotated code might be reached at runtime. We should be extremely clear in our documentation that this is not the intended use.
+
+### Validating correctness of feature guards
+
+Better would be for the analyzer to validate that the implementation of a feature check includes a check for `IsDynamicCodeSupported`. This could be done using the analyzer infrastructure we have in place without much cost for simple cases:
+
```

Comment:
There might be a need to do this type of analysis even in ILLink/ILCompiler, at least eventually. Especially if we start relying on `FeatureGuard` attributes in those tools. I don't think it's valid to rely on the analyzer for validating correctness of an app (even if for just a small portion of it). The one advantage here would be that the `FeatureGuard` attribute is introduced along side the corresponding analyzer support, but still.